### PR TITLE
Android - Prevent pending downloads to happen when going back to home

### PR DESCRIPTION
**Pull request by Cristo86**
Downloads shouldn't start when going back to home (Android)

**Test Plan**

1. Open the app
2. Enter a Domain
3. Before it loads all, expect so see one object loaded and press back immediately
4. Domains list should appear
5. Enter again to the same Domain

**Verify**

- It was experienced in previous versions that in some cases, domains list loading was slow after coming back from Interface (~15s). It shouldn't be that slow now (consider a 3s as max with decent bandwidth).
- Be sure we don't get stuck without downloading (download broken). If we try to enter again the domain after coming back, elements should resume downloading.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10771/HighFidelity-Beta-PR13431-3837952.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10771/HighFidelity-Beta-PR13431-3837952.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10771/HighFidelity-Beta-Interface-PR13431-3837952.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10771/HighFidelity-Beta-PR13431-3837952.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10771/HighFidelity-Beta-Interface-PR13431-3837952.exe

**Comment by Cristo86** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10793/HighFidelity-Beta-PR13431-3837952.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10793/HighFidelity-Beta-PR13431-3837952.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10793/HighFidelity-Beta-Interface-PR13431-3837952.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10793/HighFidelity-Beta-PR13431-3837952.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10793/HighFidelity-Beta-Interface-PR13431-3837952.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11071/HighFidelity-Beta-PR13431-6b5e570.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11071/HighFidelity-Beta-PR13431-6b5e570.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11071/HighFidelity-Beta-Interface-PR13431-6b5e570.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11071/HighFidelity-Beta-PR13431-6b5e570.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11071/HighFidelity-Beta-Interface-PR13431-6b5e570.exe

File: libraries/networking/src/ResourceCache.cpp [hunk 1]
```diff
@@ -95,6 +95,27 @@ void ResourceCacheSharedItems::removeRequest(QWeakPointer<Resource> resource) {
...
+    if (!resource) return;
+    if (ResourceCache::getRequestsActive() >= ResourceCache::getRequestLimit()) {
+        // wait until a slot becomes available
+        appendPendingRequest(resource);
+        return;
+    }
+    // ResourceCache::attemptRequest code
+    ResourceCache::incRequestsActive();
+    appendActiveRequest(resource);
+    resource->makeRequest();
```
[the remark this answers is not in the document]
Comment (Cristo86) [reply libraries/networking/src/ResourceCache.cpp]: I thought I needed a mechanism to make it start getting requests again. In tests not doing it still worked (only setting pause to false), as in those cases it seems that there always was a resource to get, starting the chain again.
We may move forward without that code.

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -8305,6 +8305,8 @@ void Application::enterBackground() {
     if (getActiveDisplayPlugin()->isActive()) {
         getActiveDisplayPlugin()->deactivate();
     }
+    DependencyManager::get<ResourceCacheSharedItems>()->pausePendingRequests();
+    clearDomainOctreeDetails();
```
[the remark this answers is not in the document]
Comment (Cristo86) [reply interface/src/Application.cpp]: clearDomainOctreeDetails prevented the following bug:
- enter domain
- exit
- enter same domain

< avatar starts falling and when the floor it was intended to step on loads, it's too late and appears over our head >
Disconnecting seems to give it more time then. (Samuel commented something similar is happening in Desktop/master? Maybe a different fix would be available).

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11121/HighFidelity-Beta-PR13431-736fc41.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11121/HighFidelity-Beta-PR13431-736fc41.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11121/HighFidelity-Beta-Interface-PR13431-736fc41.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11133/HighFidelity-Beta-PR13431-a878ad5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11133/HighFidelity-Beta-PR13431-a878ad5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11133/HighFidelity-Beta-Interface-PR13431-a878ad5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11133/HighFidelity-Beta-PR13431-a878ad5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11133/HighFidelity-Beta-Interface-PR13431-a878ad5.exe

File: interface/src/Application.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (birarda) [reply interface/src/Application.cpp]: Do we want to just fully disconnect from the domain here as opposed to just clearing the Octree on the client?

[the remark this answers is not in the document]
Comment (Cristo86) [reply interface/src/Application.cpp]: I've tried with `softReset` and `hardReset` from DomainHandler (I had to make this public for the test) (softReset calls ::disconnect) and references to _pendingRequests were still valid.

I think other strong reference to those objects (other than the entitytree) retained them but I couldn't find it.

After all that, we went on doing the pause/resume methods to at least stop next downloads from happening.

clearDomainOctreeDetails seems to make it possible to trigger the retrieval of domain objects (models, textures, etc..) again after resume (this resume method only switches the paused flag). If you have any better option in mind, we can try it.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11179/HighFidelity-Beta-PR13431-be8351d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11179/HighFidelity-Beta-PR13431-be8351d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11179/HighFidelity-Beta-Interface-PR13431-be8351d.dmg

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11274/HighFidelity-Beta-PR13431-be8351d.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11274/HighFidelity-Beta-PR13431-be8351d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11274/HighFidelity-Beta-Interface-PR13431-be8351d.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Excellent load times.

**Pull request merged**